File: abc273/e.py
class TreeNode:
    def __init__(self,parent:int,value:int):
        self.parent = parent
        self.value = value
        self.children = []
    
    def add_children(self,c):
        self.children.append(c)
    


class Tree:
    def __init__(self):
        self.tree:list[TreeNode] = []
    

    def insert_node(self,value,parent):
        n = TreeNode(parent, value)
        self.tree.append(n)
        inserted_node_idx = len(self.tree) - 1
        self.tree[parent].add_children(inserted_node_idx)
        return inserted_node_idx

if __name__ == "__main__":
    Q = int(input())
    t = Tree()
    
    note = {}
    ans = []
    cur = t.insert_node(-1, 0)
    for i in range(Q):
        query = input().split()
        q = query[0]
        if q == "ADD":
            x = int(query[1])
            cur = t.insert_node(x, cur)
        if q == "DELETE":
            cur = t.tree[cur].parent
        if q == "SAVE":
            y = int(query[1])
            note[y] = cur
        if q == "LOAD":
            z = int(query[1])
            if z not in note:
                cur = 0
            else:
                cur = note[z]
        ans.append(t.tree[cur].value)
    
    print(*ans)

File: abc273/e.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;

class TreeNode
{
private:
    int parent;
    vector<int> children;
    ll value;

public:
    TreeNode(int p, ll v)
    {
        parent = p;
        value = v;
    }
    void add_children(int c)
    {
        children.push_back(c);
    }
    int get_parent()
    {
        return parent;
    }
    ll get_value()
    {
        return value;
    }
};

int main()
{
    int Q;
    cin >> Q;
    map<int, int> note;
    vector<TreeNode> tree;
    tree.push_back(TreeNode(0, -1));
    vector<ll> ans;
    int cur_tail_A = 0;
    for (int i = 0; i < Q; i++)
    {
        string s;
        cin >> s;
        if (s == "ADD")
        {
            ll x;
            cin >> x;
            tree.push_back(TreeNode(cur_tail_A, x));
            tree[cur_tail_A].add_children(tree.size() - 1);
            cur_tail_A = tree.size() - 1;
        }
        if (s == "DELETE")
        {
            cur_tail_A = tree[cur_tail_A].get_parent();
        }
        if (s == "SAVE")
        {
            int y;
            cin >> y;
            note[y] = cur_tail_A;
        }
        if (s == "LOAD")
        {
            int z;
            cin >> z;
            cur_tail_A = note[z];
        }
        ans.push_back(tree[cur_tail_A].get_value());
    }
    for (auto a : ans)
    {
        cout << a << " ";
    }
    cout << endl;
}